### PR TITLE
Ledger Live new API updates

### PR DESCRIPTION
**Pull request by r-czajkowski**
This PR moves the implementation of the Ledger Live Bitcoin provider and connector from Orangekit to the Acre dapp. We decided to keep this implementation on the Acre side because there are many implementation details specified for Acre. Here we also use new functions from the new Ledger Live Wallet API.

### CODE REVIEW

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:70]: We can move it to the constructor and set the correct name based on the network. But let's confirm it with the Ledger team first.

File: dapp/manifests/ledger-live/ledger-live-manifest-development.json
Comment (r-czajkowski) [reply dapp/manifests/ledger-live/ledger-live-manifest-development.json:27]: I think we can remove:
- `message.sign` - we want to sign all messages via Acre module,
- `transaction.sign`,
- `transaction.signAndBroadcast` - we want to send transactions via Acre module but it doesn't work at the moment.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:135]: In my opinion, there is no need to install additional external lib to support this type in one place. In dapp we use native BigInts. The Ledger Live converts it to string under the hood anyway. We had a conversation about it some time ago please see https://github.com/thesis/acre/pull/217#discussion_r1486375599.

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts:36]: Let's update once we get the final version of this PR https://github.com/thesis/orangekit/pull/143.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:269]: See https://github.com/thesis/acre/pull/780#discussion_r1822302973

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:287]: We must implement the `signMessage` function from the `OrangeKitBitcoinWalletProvider`.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:306]: I'll update once we merge this PR https://github.com/thesis/orangekit/pull/144.

File: dapp/manifests/ledger-live/ledger-live-manifest-development.json
Comment (r-czajkowski) [reply dapp/manifests/ledger-live/ledger-live-manifest-development.json]: aaee3cc

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 6ba9a9f

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:185]: Do we need this workaround? Can't we wait for improvements from the Ledger team?

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: a3cd8a4

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 57486c0

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:185]: Yeah, let's wait. This will be our plan B.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:181]: So `bignumber.js` will be installed anyway as an indirect dependency through `@ledgerhq/wallet-api-client.` We just avoid listing it in the `package.json`.

I don't like ignoring the error, as we're losing the advantage of the type safety TS provides.
We assume that the external library performs the conversion, but if the implementation in the Ledger library changes, we will only find out at runtime.

Comment (nkuba) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:181]: Instead of adding the helper function I initially proposed, we could convert the value inline.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:292]: Ok, so `signMessage` from `OrangeKitBitcoinWalletProvider` is a general blind signing function, the Acre module supports only clear signing.
Let's implement `signMessage` to use `walletApiClient.message.sign` to match the `OrangeKitBitcoinWalletProvider` interface.
And then we have two clear signing functions for the custom Acre module: `signSignInMessage` and `signWithdrawMessage`.

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (nkuba) [reply sdk/src/lib/bitcoin/providers/provider.ts:36]: So, how about we rename the interface to `AcreBitcoinProvider`?

Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts]: 289aa1f

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 289aa1f

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 289aa1f

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: f118aaf

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts]: 2aec262

File: sdk/test/lib/redeemer-proxy.test.ts
Comment (nkuba) [inline sdk/test/lib/redeemer-proxy.test.ts:129]: Can we add tests that cover a provider that implements `signWithdrawMessage` and the one that doesn't?

File: sdk/test/utils/mock-orangekit.ts
Comment (nkuba) [inline sdk/test/utils/mock-orangekit.ts:43]:
```suggestion
    bitcoinSignMessageFn: BitcoinSignSafeTransactionMessageFn,
```

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 2d80037

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:266]: 139efb8

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts]: 139efb8

File: sdk/src/lib/redeemer-proxy.ts
Comment (r-czajkowski) [reply sdk/src/lib/redeemer-proxy.ts]: f3f4d9a

File: sdk/test/utils/mock-orangekit.ts
Comment (r-czajkowski) [reply sdk/test/utils/mock-orangekit.ts]: 8350245

File: sdk/test/lib/redeemer-proxy.test.ts
Comment (r-czajkowski) [reply sdk/test/lib/redeemer-proxy.test.ts]: 7ff32cb

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:286]: > Let's implement `signMessage` to use `walletApiClient.message.sign` to match the `OrangeKitBitcoinWalletProvider` interface.

The `walletApiClient.message.sign` always signs the message using the fresh address so this may be confusing. Another thing is that we use the `useSignMessage` hook from `wagmi` which calls the `signMessage` function from the `bitcoinProvider` see https://github.com/thesis/orangekit/blob/main/typescript/react/src/wallet/client.ts#L32. Since the Acre dapp currently requires 2 types of messages: SIWW message (it's just string) and withdraw, and we have the custom bitcoin provider used only in Acre dapp I assumed the `signMessage` signs the message using the Acre module. Also, with the current implementation of Acre module, we can sign any message using the Acre module and `SignIn` type.

Comment (nkuba) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:286]: You raised important points. We shouldn't block this PR to solve this matter since it requires work on the OrangeKit and Ledger API sides.

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [inline dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:50]: Let's rename it to match the provider name.
On the other hand, do we need to export it? Can't it be simply named `Options`?

```suggestion
export type AcreLedgerLiveBitcoinProviderOptions = {
```

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [inline dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:156]: Shouldn't we call `#getAddress` here to return the address at 0 index?

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (nkuba) [inline dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:164]:
```suggestion
   * @returns The transaction hash.
```

File: dapp/src/utils/orangekit/ledger-live/tests/bitcoin-provider.test.ts
Comment (nkuba) [inline dapp/src/utils/orangekit/ledger-live/tests/bitcoin-provider.test.ts:397]: Shouldn't this test return the address at 0-index?

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts:50]: We use this type in `connector.ts`.

File: dapp/src/utils/orangekit/ledger-live/tests/bitcoin-provider.test.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/tests/bitcoin-provider.test.ts:406]: bb447b2

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 6cde85b

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 2164f13

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts:37]: 4a7ac74

File: dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts
Comment (r-czajkowski) [reply dapp/src/utils/orangekit/ledger-live/bitcoin-provider.ts]: 5de537c

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (nkuba) [inline sdk/src/lib/bitcoin/providers/provider.ts:40]: We're missing a line break before `@returns`.

**Review by nkuba — APPROVED** [review]
I left one non-blocking comment, that can be addressed in [another PR.](https://github.com/thesis/acre/pull/708).

File: sdk/src/lib/bitcoin/providers/provider.ts
Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts:40]: 🤦

Comment (r-czajkowski) [reply sdk/src/lib/bitcoin/providers/provider.ts:40]: Fixed in be9fa836dd5dbad164aa1d13b233ce803f529f17.

**Pull request merged**
